util/filesort/filesort.go

```diff
@@ -0,0 +1,508 @@
+	)
+	n, err = fs.fds[index].Read(head)
+	if err == io.EOF {
+		return nil, nil
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != 8 {
+		return nil, errors.New("incorrect header")
+	}
+	rowSize := int(binary.BigEndian.Uint64(head))
+
+	rowBytes := make([]byte, rowSize)
+
+	n, err = fs.fds[index].Read(rowBytes)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if n != rowSize {
+		return nil, errors.New("incorrect row")
+	}
+
+	dcod, err = codec.Decode(rowBytes, fs.keySize+fs.valSize+1)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &comparableRow{
+		key:    dcod[:fs.keySize],
+		val:    dcod[fs.keySize : fs.keySize+fs.valSize],
+		handle: dcod[fs.keySize+fs.valSize:][0].GetInt64(),
+	}, nil
+}
+
+func (fs *FileSorter) openAllFiles() error {
+	for _, fname := range fs.files {
+		fd, err := os.Open(fname)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		fs.fds = append(fs.fds, fd)
+	}
+	return nil
+}
+
+func (fs *FileSorter) closeAllFiles() error {
+	for _, fd := range fs.fds {
+		err := fd.Close()
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	err := os.RemoveAll(fs.tmpDir)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Close terminates the output process and discards all remaining data.
+func (fs *FileSorter) Close() error {
+	if fs.closed {
+		return errors.New("FileSorter has been closed")
+	}
+	fs.buf = fs.buf[:0]
+	err := fs.closeAllFiles()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	fs.closed = true
+	return nil
+}
+
+// Output gets the next sorted row.
+func (fs *FileSorter) Output() (key []types.Datum, val []types.Datum, handle int64, err error) {
+	if fs.closed {
+		return nil, nil, 0, errors.New("FileSorter has been closed")
+	}
+	if len(fs.files) == 0 {
+		// No external files generated.
+		// Perform full in-memory sort directly.
+		return fs.internalSort()
+	}
+	return fs.externalSort()
+}
+
+// Perform full in-memory sort.
+func (fs *FileSorter) internalSort() (key []types.Datum, val []types.Datum, handle int64, err error) {
+	if !fs.fetched {
+		sort.Sort(fs)
+		if fs.err != nil {
+			return nil, nil, 0, errors.Trace(fs.err)
+		}
+		fs.fetched = true
+	}
+	if fs.cursor < len(fs.buf) {
+		r := fs.buf[fs.cursor]
+		fs.cursor++
+		return r.key, r.val, r.handle, nil
+	}
+	return nil, nil, 0, nil
+}
+
+// Perform external file sort.
+func (fs *FileSorter) externalSort() (key []types.Datum, val []types.Datum, handle int64, err error) {
+	if !fs.fetched {
+		if len(fs.buf) > 0 {
+			err = fs.flushToFile()
+			if err != nil {
+				return nil, nil, 0, errors.Trace(err)
+			}
+		}
+
+		heap.Init(fs.rowHeap)
+		if fs.rowHeap.err != nil {
+			return nil, nil, 0, errors.Trace(fs.rowHeap.err)
+		}
+
+		err = fs.openAllFiles()
+		if err != nil {
+			return nil, nil, 0, errors.Trace(err)
+		}
+
+		for id := range fs.fds {
+			row, err := fs.fetchNextRow(id)
+			if row == nil {
+				if err != nil {
```

Comment:
Actually `fetchNextRow` is kind of special. `err == nil` has two cases: 1) successfully fetched a new row. 2) met the `EOF`. If we choose to judge the error first, we might miss the second case. Since this piece of code is about heap initialization, it should not meet the `EOF` at here (All files should not be empty at the first place). So when `err == nil` and `row == nil`, we need to complain.